controllers/openstackvmset_controller.go

```diff
@@ -1140,83 +852,770 @@ func (r *OpenStackVMSetReconciler) vmCreateInstance(
+
+	// Generate existing host NetworkData next
+	for hostname, actualStatus := range instance.Status.DeepCopy().VMHosts {
+
+		if !common.StringInSlice(hostname, newVMs) && hostname != "" {
+			if err := generateNetworkData(instance, &actualStatus); err != nil {
+				return vmDetails, err
+			}
+			currentStatus := instance.Status.VMHosts[hostname]
+			if !reflect.DeepEqual(currentStatus, actualStatus) {
+				instance.Status.VMHosts[hostname] = actualStatus
+				common.LogForObject(
+					r,
+					fmt.Sprintf("Changed Host NetStatus, updating CR status current - %s: %v / new: %v",
+						hostname,
+						actualStatus,
+						diff.ObjectReflectDiff(currentStatus, actualStatus)),
+					instance,
+				)
+			}
+		}
+	}
+
+	return vmDetails, nil
+}
+
+//
+//   Create the VM objects
+//
+func (r *OpenStackVMSetReconciler) createVMs(
+	instance *ospdirectorv1beta1.OpenStackVMSet,
+	cond *ospdirectorv1beta1.Condition,
+	envVars map[string]common.EnvSetter,
+	osNetBindings map[string]ospdirectorv1beta1.AttachType,
+	baseImageName string,
+	vmDetails map[string]ospdirectorv1beta1.Host,
+) (ctrl.Result, error) {
+	if instance.Status.BaseImageDVReady {
+		for _, ctl := range vmDetails {
+			// Add chosen baseImageName to controller details, then create the VM instance
+			ctl.BaseImageName = baseImageName
+			err := r.vmCreateInstance(instance, cond, envVars, &ctl, osNetBindings)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+
+		// Calculate provisioning status
+		readyCount := 0
+		for _, host := range instance.Status.VMHosts {
+			if host.ProvisioningState == ospdirectorv1beta1.VMSetCondTypeProvisioned {
+				readyCount++
+			}
+		}
+		instance.Status.ProvisioningStatus.ReadyCount = readyCount
+
+		switch readyCount := instance.Status.ProvisioningStatus.ReadyCount; {
+		case readyCount == instance.Spec.VMCount && readyCount == 0:
+			cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.VMSetCondTypeEmpty)
+			cond.Reason = ospdirectorv1beta1.ConditionReason(ospdirectorv1beta1.VMSetCondReasonVirtualMachineCountZero)
+			cond.Message = "No VirtualMachines have been requested"
+		case readyCount == instance.Spec.VMCount:
+			cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.VMSetCondTypeProvisioned)
+			cond.Reason = ospdirectorv1beta1.ConditionReason(ospdirectorv1beta1.VMSetCondReasonVirtualMachineProvisioned)
+			cond.Message = "All requested VirtualMachines have been provisioned"
+		case readyCount < instance.Spec.VMCount:
+			cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.VMSetCondTypeProvisioning)
+			cond.Reason = ospdirectorv1beta1.ConditionReason(ospdirectorv1beta1.VMSetCondReasonVirtualMachineProvisioning)
+			cond.Message = "Provisioning of VirtualMachines in progress"
+		default:
+			cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.VMSetCondTypeDeprovisioning)
+			cond.Reason = ospdirectorv1beta1.ConditionReason(ospdirectorv1beta1.VMSetCondReasonVirtualMachineProvisioning)
```

Comment:
Should we have a separate condition reason for deprovisioning?  I'm fine with the way you have it, but I thought I'd just ask.

Reply:
yes makes sense. I added one